File: lib/spark/tools/seed/sparkwrap.c
#define _GNU_SOURCE
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <libgen.h>
#include <limits.h>
#include <errno.h>

int main(int argc, char *argv[]) {
    char path[PATH_MAX];
    char exe_path[PATH_MAX];
    char *dir;
    ssize_t len;

    // Get directory from environment or executable path
    dir = getenv("SPARK_SEED_DIR");
    if (!dir) {
        len = readlink("/proc/self/exe", exe_path, sizeof(exe_path) - 1);
        if (len == -1) {
            perror("Failed to get executable path");
            return 1;
        }
        exe_path[len] = '\0';
        dir = dirname(exe_path);
    }

    // Build path to real binary
    if (snprintf(path, sizeof(path), "%s/zig-out/bin/seed", dir) >= sizeof(path)) {
        fprintf(stderr, "Path too long\n");
        return 1;
    }

    // Execute the real binary with raw arguments
    execv(path, argv);
    
    // If execv fails, print error and exit
    perror("Failed to execute seed manager");
    return 1;
}

File: lib/spark/tools/seed/sparkwrap.zig
const std = @import("std");
const os = std.os;
const fs = std.fs;

pub fn main() !void {
    // Get current executable path
    var buf: [fs.MAX_PATH_BYTES]u8 = undefined;
    const self_path = try fs.selfExePath(&buf);
    const dir = fs.path.dirname(self_path) orelse return error.NoPath;
    
    // Build path to real binary
    var path_buf: [fs.MAX_PATH_BYTES]u8 = undefined;
    const real_path = try std.fmt.bufPrint(&path_buf, "{s}/zig-out/bin/seed", .{dir});
    
    // Execute the real binary, bypassing shell interpretation
    const err = os.linux.syscall3(
        .execve,
        @intFromPtr(real_path.ptr),
        @intFromPtr(os.argv.ptr),
        @intFromPtr(@as(?[*]const ?[*:0]u8, null))
    );
    
    switch (err) {
        0 => {},
        else => return error.ExecveFailed,
    }
}
